src/main/scala/decider/Z3ProverAPI.scala

```diff
@@ -0,0 +1,439 @@
+    prover.add(negatedGoal)
+    val startTime = System.currentTimeMillis()
+    val res = prover.check()
+    val endTime = System.currentTimeMillis()
+    val result = res == Status.UNSATISFIABLE
+    pop()
+
+    if (!result) {
+      retrieveAndSaveModel()
+    }
+
+    (result, endTime - startTime)
+  }
+
+  def saturate(data: Option[Config.ProverStateSaturationTimeout]): Unit = {
+    endPreamblePhase()
+    data match {
+      case Some(Config.ProverStateSaturationTimeout(timeout, comment)) => saturate(timeout, comment)
+      case None => /* Don't do anything */
+    }
+  }
+
+  def saturate(timeout: Int, comment: String): Unit = {
+    endPreamblePhase()
+    setTimeout(Some(timeout))
+    prover.check()
+  }
+
+  protected def retrieveAndSaveModel(): Unit = {
+    if (Verifier.config.counterexample.toOption.isDefined) {
+      val model = prover.getModel
+      lastModel = model
+    }
+  }
+
+  protected def assertUsingSoftConstraints(goal: Term): (Boolean, Long) = {
+    endPreamblePhase()
+    val guard = fresh("grd", Nil, sorts.Bool)
+    val guardApp = App(guard, Nil)
+    val goalImplication = Implies(guardApp, goal)
+
+    prover.add(termConverter.convertTerm(goalImplication).asInstanceOf[BoolExpr])
+
+    val startTime = System.currentTimeMillis()
+    val res = prover.check(termConverter.convertTerm(guardApp))
+    val endTime = System.currentTimeMillis()
+    val result = res == Status.UNSATISFIABLE
+    if (!result) {
+      retrieveAndSaveModel()
+    }
+
+    (result, endTime - startTime)
+  }
+
+  def check(timeout: Option[Int] = None): Result = {
+    endPreamblePhase()
+    setTimeout(timeout)
+
+    val res = prover.check()
+
+    res match {
+      case Status.SATISFIABLE => Sat
+      case Status.UNSATISFIABLE => Unsat
+      case Status.UNKNOWN=> Unknown
+    }
+  }
+
+  def endPreamblePhase() =  {
+    if (!preamblePhaseOver) {
+      preamblePhaseOver = true
+
+      val merged = emittedPreambleString.mkString("\n")
+      val parsed = ctx.parseSMTLIB2String(merged, emittedSortSymbols.toArray, emittedSorts.toArray, emittedFuncSymbols.toArray, emittedFuncs.toArray)
+      prover.add(parsed: _*)
+      prover.add(preambleAssumes.toSeq : _*)
+      preambleAssumes.clear()
+    }
+  }
+
+  def statistics(): Map[String, String] = {
+    val statistics = prover.getStatistics
+    val result = mutable.HashMap[String, String]()
+    for (e <- statistics.getEntries()) {
+      result.update(e.Key, e.getValueString)
+    }
+    ListMap.from(result)
+  }
+
+  def comment(str: String): Unit = {
+    if (!preamblePhaseOver && str == "End preamble")
+      endPreamblePhase()
+  }
+
+  def fresh(name: String, argSorts: Seq[Sort], resultSort: Sort): Fun = {
+    val id = identifierFactory.fresh(name)
+    val fun = Fun(id, argSorts, resultSort)
+    val decl = FunctionDecl(fun)
+
+    termConverter.convert(decl)
+
+    fun
+  }
+
+  def declare(decl: Decl): Unit = {
+    // just creating the term is enough.
```

Comment:
What is this comment referring to? The declaration is still added to `emitted*`, no?

Reply:
When calling ctx.mkFuncDecl, the declaration is added to Z3's current state, without us having to explicitly call another function (e.g., ctx.declareFunc(ctx.mkFuncDecl(...)). That's what I meant there, will clarify.